Basics of Flask:

1. Create a Flask app that displays "Hello, World!" on the homepage.

In [1]:
from flask import Flask

app = Flask(__name__)

@app.route("/")
def hello_world():
    return "<h1>Hello, World!</h1>"

if __name__=="__main__":
    app.run(host="0.0.0.0",port=5001)
# https://red-hairdresser-cllvt.pwskills.app:5001/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5001
 * Running on http://192.168.105.188:5001
Press CTRL+C to quit
127.0.0.1 - - [04/Nov/2023 23:46:30] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2023 23:46:30] "GET /favicon.ico HTTP/1.1" 404 -


2. Build a Flask app with static HTML pages and navigate between them.

# Templates : Folder
<!-- Index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Home</title>
</head>
<body>
    <h1>Welcome to the Home Page</h1>
    <a href="/page1">Go to Page 1</a>
    <a href="/page2">Go to Page 2</a>
</body>
</html>

<!-- page1.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Home</title>
</head>
<body>
    <h1>Welcome to the Page1</h1>
    <a href="/">Go to Home Page</a>
</body>
</html>

<!-- page2.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Home</title>
</head>
<body>
    <h1>Welcome to the Page2</h1>
    <a href="/">Go to Home Page</a>
</body>
</html>

# Outside templates folder
<!-- app.py -->
from flask import Flask, render_template

app = Flask(__name__)

@app.route('/')
def home():
    return render_template('index.html')

@app.route('/page1')
def page1():
    return render_template('page1.html')

@app.route('/page2')
def page2():
    return render_template('page2.html')

if __name__ == '__main__':
    app.run(host="0.0.0.0",port=5002)
    # https://red-hairdresser-cllvt.pwskills.app:5000/

3. Develop a Flask app that uses URL parameters to display dynamic content.

<!-- dynamic.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Dynamic Content</title>
</head>
<body>
    <h1>Hello, {{ name }}!</h1>
</body>
</html>

<!-- app.py -->
from flask import Flask, render_template

app = Flask(__name__)

@app.route('/greet/<name>')
def greet(name):
    return render_template('dynamic.html', name=name)

if __name__=="__main__":
    app.run(host="0.0.0.0",port=5004)
    # https://mango-architect-isvlc.pwskills.app:5004/greet/Ganesha
    # https://mango-architect-isvlc.pwskills.app:5004/greet/Ram

4. Create a Flask app with a form that accepts user input and displays it.

<!-- index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Form Example</title>
</head>
<body>
    <h1>Submit a Form</h1>
    <form action="/check_password" method="post">
        <label for="username">Username:</label>
        <input type="text" id="username" name="username" required><br><br>

        <label for="password">Password:</label>
        <input type="password" id="password" name="password" required><br><br>

        <input type="submit" value="ok">
    </form>
</body>
</html>



<!-- userform.py -->
from flask import Flask, request, jsonify, render_template

app = Flask(__name__)

@app.route('/') # url and port number
def show_form():
    return render_template('index.html')
# render_template : show html file

# API : /check_password
@app.route('/check_password', methods=['POST'])
def check_password():
    name = request.form.get('username')
    password = request.form.get('password')
    print({name}, {password})
    return "username and password received"

if __name__ == "__main__":
    app.run(host="0.0.0.0",port="5002")
# only 5000 and 8000 open ports

5. Implement user sessions in a Flask app to store and display user-specific data.

<!-- Index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>User Session Example</title>
</head>
<body>
    <h1>Welcome</h1>
    {% if user_data %}
    <p>Hello, {{ user_data.username }}!</p>
    <a href="/logout">Logout</a>
    {% else %}
    <form method="post" action="/login">
        <label for="username">Username:</label>
        <input type="text" name="username" id="username">
        <input type="submit" value="Login">
    </form>
    {% endif %}
</body>
</html>

<!-- app.py -->
from flask import Flask, render_template, request, session
from flask_session import Session

app = Flask(__name__)

# Configure session to use cookies as the storage backend
app.config["SESSION_TYPE"] = "filesystem"  # You can change this to other storage options
app.config['PERMANENT_SESSION_LIFETIME'] = 10  # 10 seconds
Session(app)

# Set a secret key to secure your session data
app.secret_key = 'your_secret_key_here'

@app.route('/')
def index():
    # Get user-specific data from the session (if available)
    user_data = session.get('user_data', {})
    return render_template('index.html', user_data=user_data)

@app.route('/login', methods=['POST'])
def login():
    # Simulate user login - store user-specific data in the session
    username = request.form.get('username')
    user_data = {
        'username': username,
        # Add more user-specific data here
    }
    session['user_data'] = user_data
    return 'Logged in as ' + username

@app.route('/logout')
def logout():
    # Clear user-specific data from the session
    session.pop('user_data', None)
    return 'Logged out'

if __name__=="__main__":
    app.run(host="0.0.0.0")

**Intermediate Flask Topics:**

6. Build a Flask app that allows users to upload files and display them on the website.

create templates folder
create uploads folder
create app.py

<!-- upload.html -->
<!DOCTYPE html>
<html>
<head>
    <title>File Upload</title>
</head>
<body>
    <h1>Upload a File</h1>
    <form method="POST" action="/upload" enctype="multipart/form-data">
        <input type="file" name="file">
        <input type="submit" value="Upload">
    </form>
    <hr>
    {% if file_url %}
        <h2>Uploaded File:</h2>
        <img src="{{ file_url }}" alt="Uploaded File">
    {% endif %}
</body>
</html>


<!-- app.py -->
import os
from flask import Flask, render_template, request, send_from_directory

app = Flask(__name__)

# Configure the upload folder
app.config['UPLOAD_FOLDER'] = 'uploads'
app.config['ALLOWED_EXTENSIONS'] = {'txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif'}

def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in app.config['ALLOWED_EXTENSIONS']

@app.route('/')
def upload_form():
    return render_template('upload.html')

@app.route('/upload', methods=['POST'])
def upload_file():
    if 'file' not in request.files:
        return 'No file part'

    file = request.files['file']

    if file.filename == '':
        return 'No selected file'

    if file and allowed_file(file.filename):
        file.save(os.path.join(app.config['UPLOAD_FOLDER'], file.filename))
        file_url = f"/uploads/{file.filename}"
        return render_template('upload.html', file_url=file_url)

    return 'File type not allowed'

@app.route('/uploads/<filename>')
def uploaded_file(filename):
    return send_from_directory(app.config['UPLOAD_FOLDER'], filename)

if __name__=="__main__":
    app.run(host="0.0.0.0")

7. Integrate a SQLite database with Flask to perform CRUD operations on a list of items.

<!-- Index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Items List</title>
</head>
<body>
    <h1>Items List</h1>
    <ul>
        {% for item in items %}
            <li>
                {{ item.name }}
                <a href="/delete_item/{{ item.id }}">Delete</a>
            </li>
        {% endfor %}
    </ul>
    <form method="post" action="/add_item">
        <input type="text" name="item_name" placeholder="Add an item" required>
        <input type="submit" value="Add">
    </form>
</body>
</html>


<!-- app.py -->
from flask import Flask, render_template, request, redirect, url_for
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///items.db'  # SQLite database file
db = SQLAlchemy(app)

class Item(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)

@app.route('/')
def index():
    items = Item.query.all()
    return render_template('index.html', items=items)

@app.route('/add_item', methods=['POST'])
def add_item():
    item_name = request.form.get('item_name')
    if item_name:
        new_item = Item(name=item_name)
        db.session.add(new_item)
        db.session.commit()
    return redirect(url_for('index'))

@app.route('/delete_item/<int:item_id>')
def delete_item(item_id):
    item = Item.query.get(item_id)
    if item:
        db.session.delete(item)
        db.session.commit()
    return redirect(url_for('index'))

if __name__=="__main__":
    with app.app_context():
        db.create_all()
    app.run(host="0.0.0.0")

8. Implement user authentication and registration in a Flask app using Flask-Login.

<!-- login.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Login</title>
</head>
<body>
    <h1>Login</h1>
    <form method="post">
        <label for="username">Username:</label>
        <input type="text" name="username" id="username" required>
        <br>
        <label for="password">Password:</label>
        <input type="password" name="password" id="password" required>
        <br>
        <input type="submit" value="Log In">
    </form>
</body>
</html>

<!-- register.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Register</title>
</head>
<body>
    <h1>Register</h1>
    <form method="post">
        <label for="username">Username:</label>
        <input type="text" name="username" id="username" required>
        <br>
        <label for="password">Password:</label>
        <input type="password" name="password" id="password" required>
        <br>
        <input type="submit" value="Register">
    </form>
</body>
</html>

<!-- dashboard.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Dashboard</title>
</head>
<body>
    <h1>Dashboard</h1>
    <p>Welcome, {{ user.username }}!</p>
    <a href="/logout">Logout</a>
</body>
</html>

<!-- app.py -->
from flask import Flask, render_template, request, redirect, url_for, flash
from flask_login import LoginManager, UserMixin, login_user, login_required, logout_user, current_user

app = Flask(__name__)
app.secret_key = 'your_secret_key_here'

# Create in-memory storage for simplicity; in a real app, use a database.
users = {'user1': {'password': 'password1'}, 'user2': {'password': 'password2'}}

login_manager = LoginManager(app)
login_manager.login_view = "login"

class User(UserMixin):
    pass

@login_manager.user_loader
def load_user(username):
    if username in users:
        user = User()
        user.id = username
        return user

@app.route('/')
@login_required
def dashboard():
    return render_template('dashboard.html', user=current_user)

@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        username = request.form['username']
        password = request.form['password']
        if username in users and users[username]['password'] == password:
            user = User()
            user.id = username
            login_user(user)
            flash('Logged in successfully', 'success')
            return redirect(url_for('dashboard'))
        else:
            flash('Login failed. Please check your username and password.', 'error')
    return render_template('login.html')

@app.route('/register', methods=['GET', 'POST'])
def register():
    if request.method == 'POST':
        username = request.form['username']
        password = request.form['password']
        if username not in users:
            users[username] = {'password': password}
            flash('Registration successful. Please log in.', 'success')
            return redirect(url_for('login'))
        else:
            flash('Registration failed. Username already exists.', 'error')
    return render_template('register.html')

@app.route('/logout')
@login_required
def logout():
    logout_user()
    flash('Logged out successfully', 'success')
    return redirect(url_for('login'))

if __name__=="__main__":
    app.run(host="0.0.0.0")

9. Create a RESTful API using Flask to perform CRUD operations on resources like books or movies.

from flask import Flask, request, jsonify
from data import books  # Import your data (books, movies, etc.)

app = Flask(__name__)

# Sample data (in data.py) - Replace this with your database or data source
# Example data for books:
# books = [
#     {"id": 1, "title": "Book 1", "author": "Author 1"},
#     {"id": 2, "title": "Book 2", "author": "Author 2"},
# ]

@app.route('/api/books', methods=['GET'])
def get_books():
    return jsonify(books)

@app.route('/api/books/<int:book_id>', methods=['GET'])
def get_book(book_id):
    book = next((b for b in books if b["id"] == book_id), None)
    if book:
        return jsonify(book)
    return "Book not found", 404

@app.route('/api/books', methods=['POST'])
def create_book():
    data = request.get_json()
    new_book = {
        "id": len(books) + 1,
        "title": data.get("title"),
        "author": data.get("author"),
    }
    books.append(new_book)
    return jsonify(new_book), 201

@app.route('/api/books/<int:book_id>', methods=['PUT'])
def update_book(book_id):
    data = request.get_json()
    book = next((b for b in books if b["id"] == book_id), None)
    if book:
        book["title"] = data.get("title")
        book["author"] = data.get("author")
        return jsonify(book)
    return "Book not found", 404

@app.route('/api/books/<int:book_id>', methods=['DELETE'])
def delete_book(book_id):
    book = next((b for b in books if b["id"] == book_id), None)
    if book:
        books.remove(book)
        return "Book deleted", 204
    return "Book not found", 404

if __name__=="__main__":
    app.run(host="0.0.0.0")

10. Design a Flask app with proper error handling for 404 and 500 errors.

<!-- template_404.html -->
<!DOCTYPE html>
<html>
<head>
    <title>404 Not Found</title>
</head>
<body>
    <h1>404 Not Found</h1>
    <p>The page you are looking for does not exist.</p>
</body>
</html>

<!-- template_500.html -->
<!DOCTYPE html>
<html>
<head>
    <title>500 Internal Server Error</title>
</head>
<body>
    <h1>500 Internal Server Error</h1>
    <p>Something went wrong on the server.</p>
</body>
</html>


<!-- app.py -->
from flask import Flask, render_template

app = Flask(__name)

# Custom error handlers for 404 and 500 errors

@app.errorhandler(404)
def not_found_error(error):
    return render_template('404.html'), 404

@app.errorhandler(500)
def internal_server_error(error):
    return render_template('500.html'), 500

# Sample route that raises a 500 error

@app.route('/error')
def trigger_error():
    # Simulate an internal server error
    1 / 0

if __name__ == '__main__':
    app.run(debug=True)

Real-time Development:

11. Create a real-time chat application using Flask-SocketIO.

<!-- chat.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Chat</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.0.1/socket.io.js"></script>
</head>
<body>
    <ul id="messages"></ul>
    <input id="input" autocomplete="off" /><button>Send</button>
    <script src="{{ url_for('static', filename='script.js') }}"></script>
</body>
</html>

<!-- script.js -->
document.addEventListener('DOMContentLoaded', () => {
    const socket = io.connect('http://' + document.domain + ':' + location.port);
    const input = document.getElementById('input');
    const messages = document.getElementById('messages');

    socket.on('message', (data) => {
        const li = document.createElement('li');
        li.innerText = data;
        messages.appendChild(li);
    });

    document.querySelector('button').onclick = () => {
        const message = input.value;
        input.value = '';
        socket.emit('message', message);
    };
});

<!-- app.py -->
from flask import Flask, render_template
from flask_socketio import SocketIO, emit

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app)

@app.route('/')
def index():
    return render_template('chat.html')

@socketio.on('message')
def handle_message(message):
    emit('message', message, broadcast=True)

if __name__=="__main__":
    app.run(host="0.0.0.0")

12. Build a Flask app that updates data in real-time using WebSocket connections.

<!-- index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Real-Time Data Update</title>
</head>
<body>
    <h1>Real-Time Data: <span id="data"></span></h1>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.0.1/socket.io.js"></script>
    <script>
        var socket = io.connect('http://' + document.domain + ':' + location.port);
        socket.on('data_updated', function(data) {
            document.getElementById('data').textContent = data;
        });
        socket.emit('update_data');
    </script>
</body>
</html>

<!-- app.py -->
from flask import Flask, render_template
from flask_socketio import SocketIO, emit
import random
import time

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app)

data = {
    'value': 0
}

@app.route('/')
def index():
    return render_template('index.html', value=data['value'])

@socketio.on('update_data')
def update_data():
    while True:
        data['value'] = random.randint(0, 100)
        emit('data_updated', data['value'], broadcast=True)
        time.sleep(2)

if __name__=="__main__":
    app.run(host="0.0.0.0")

13. Implement notifications in a Flask app using websockets to notify users of updates.

<!-- index.html -->
<!DOCTYPE html>
<html>
<head>
    <title>Real-Time Notifications</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.0.1/socket.io.js"></script>
</head>
<body>
    <h1>Real-Time Notifications</h1>
    <ul id="notifications"></ul>
    <input id="message" autocomplete="off" /><button id="send">Send Notification</button>
    <script src="{{ url_for('static', filename='script.js') }}"></script>
</body>
</html>

<!-- script.js -->
document.addEventListener('DOMContentLoaded', () => {
    const socket = io.connect('http://' + document.domain + ':' + location.port);

    socket.on('new_notification', (message) => {
        const ul = document.getElementById('notifications');
        const li = document.createElement('li');
        li.innerText = message;
        ul.appendChild(li);
    });

    const sendButton = document.getElementById('send');
    sendButton.onclick = () => {
        const messageInput = document.getElementById('message');
        const message = messageInput.value;
        socket.emit('send_notification', { message: message });
        messageInput.value = '';
    };
});


<!-- app.py -->
from flask import Flask, render_template, request, jsonify
from flask_socketio import SocketIO, emit

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app)

connected_users = set()

@app.route('/')
def index():
    return render_template('index.html')

@socketio.on('connect')
def handle_connect():
    user_id = request.sid
    connected_users.add(user_id)
    emit('user_connected', user_id, broadcast=True)

@socketio.on('disconnect')
def handle_disconnect():
    user_id = request.sid
    connected_users.remove(user_id)
    emit('user_disconnected', user_id, broadcast=True)

@app.route('/send_notification', methods=['POST'])
def send_notification():
    message = request.form['message']
    emit('new_notification', message, broadcast=True)
    return jsonify({'status': 'Message sent'})

if __name__ == '__main__':
    socketio.run(app, debug=True)
